File: game/scripts/vscripts/pve/abilities/heroes/centaur/centaur_range_attack/centaur_range_attack.lua
centaur_range_attack = class({})

function centaur_range_attack:GetCastAnimationCustom() return ACT_DOTA_ATTACK end
function centaur_range_attack:GetPlaybackRateOverride() return 0.8 end
function centaur_range_attack:GetCastPointSpeed() return 0 end


function centaur_range_attack:OnAbilityPhaseStart()
	EFX('particles/units/heroes/hero_dawnbreaker/dawnbreaker_converge.vpcf', PATTACH_ABSORIGIN_FOLLOW, self:GetCaster(), { relese = true })
	return true
end


function centaur_range_attack:OnSpellStart()
    local caster = self:GetCaster()
	local origin = caster:GetOrigin()
	local point = self:GetCursorPosition()
	local straight_direction = Direction2D(origin, point)

    
    local directions = {}
    table.insert(directions, straight_direction)
    table.insert(directions, RotatePoint(Vector(0,0,0), Vector(0,0,0) + straight_direction, -30):Normalized())
    table.insert(directions, RotatePoint(Vector(0,0,0), Vector(0,0,0) + straight_direction, 30):Normalized())

    for _,direction in ipairs(directions) do
        self:LaunchProjectile(direction)
    end
	self:PlayEffectsOnCast()
end

function centaur_range_attack:LaunchProjectile(vDirection)
    local caster = self:GetCaster()
	local origin = caster:GetOrigin()
	local projectile_speed = self:GetSpecialValueFor("projectile_speed")
	local damage = self:GetSpecialValueFor("ability_damage")
	local damage_table = {
        damage = damage,
		damage_type = DAMAGE_TYPE_MAGICAL,
	}

	CustomEntitiesLegacy:ProjectileAttack(caster, {
		tProjectile = {
			EffectName = "particles/units/heroes/hero_invoker/invoker_deafening_blast.vpcf",
			vSpawnOrigin = origin + Vector(vDirection.x * 45, vDirection.y * 45, 96),
			fDistance = self:GetSpecialValueFor("projectile_distance") ~= 0 and self:GetSpecialValueFor("projectile_distance") or self:GetCastRange(Vector(0,0,0), nil),
			fStartRadius = self:GetSpecialValueFor("hitbox"),
			Source = caster,
			vVelocity = vDirection * projectile_speed,
			UnitBehavior = PROJECTILES_NOTHING,
			TreeBehavior = PROJECTILES_NOTHING,
			WallBehavior = PROJECTILES_DESTROY,
			GroundBehavior = PROJECTILES_NOTHING,
			fGroundOffset = 0,
			UnitTest = function(_self, unit) return unit:GetUnitName() ~= "npc_dummy_unit" and not CustomEntitiesLegacy:Allies(_self.Source, unit) end,
			OnUnitHit = function(_self, unit)
				damage_table.victim = unit
				damage_table.attacker = _self.Source
				
				ApplyDamage(damage_table)
				self:PlayEffectsOnHit(unit)
			end,
			OnFinish = function(_self, pos)
				self:PlayEffectsOnFinish(pos, 'particles/units/heroes/hero_sniper/sniper_assassinate_impact_sparks.vpcf')
			end,
		}
	})
end

function centaur_range_attack:PlayEffectsOnCast()
	EmitSoundOn("Hero_Centaur.DoubleEdge.TI9", self:GetCaster())
end

function centaur_range_attack:PlayEffectsOnFinish(pos, particle_cast)
	local caster = self:GetCaster()
	EmitSoundOnLocationWithCaster(pos, "Hero_Centaur.DoubleEdge", caster)
	EmitSoundOnLocationWithCaster(pos, "Hero_Centaur.DoubleEdge.TI9_layer", caster)
	EmitSoundOnLocationWithCaster(pos, "Hero_Centaur.Gore", caster)

	local effect_cast = ParticleManager:CreateParticle(particle_cast, PATTACH_ABSORIGIN, caster)
	ParticleManager:SetParticleControl(effect_cast, 0, pos)
	ParticleManager:SetParticleControl(effect_cast, 1, pos)
	ParticleManager:SetParticleControl(effect_cast, 3, pos)
	
	ParticleManager:ReleaseParticleIndex(effect_cast)
end

function centaur_range_attack:PlayEffectsOnHit(hTarget)
	local caster = self:GetCaster()
	EmitSoundOn("Hero_Sniper.AssassinateDamage", caster)

	local particle_cast = "particles/units/heroes/hero_sniper/sniper_assassinate_impact_blood.vpcf"
	local effect_cast = ParticleManager:CreateParticle(particle_cast, PATTACH_ABSORIGIN_FOLLOW, hTarget)

	ParticleManager:SetParticleControl(effect_cast, 0, hTarget:GetOrigin())
	ParticleManager:SetParticleControl(effect_cast, 1, hTarget:GetOrigin())
	ParticleManager:ReleaseParticleIndex(effect_cast)
end

if IsClient() then require("wrappers/abilities") end
Abilities.Castpoint(centaur_range_attack)

File: src/vscripts/pve/abilities/heroes/centaur/centaur_range_attack.ts
import { CustomAbility } from "../../../../abilities/framework/custom_ability";
import { registerAbility } from "../../../../lib/dota_ts_adapter";
import { precache, resource } from "../../../../precache";
import { ProjectileBehavior } from "../../../../projectiles";
import { areUnitsAllied, direction2D, fakeAbility } from "../../../../util";

const resources = precache({
    castPoint: resource.fx("particles/units/heroes/hero_dawnbreaker/dawnbreaker_converge.vpcf"),
    projectile: resource.fx("particles/units/heroes/hero_invoker/invoker_deafening_blast.vpcf"),
    impact: resource.fx("particles/units/heroes/hero_sniper/sniper_assassinate_impact_sparks.vpcf"),
    impactBlood: resource.fx("particles/units/heroes/hero_sniper/sniper_assassinate_impact_blood.vpcf")
});

@registerAbility("centaur_range_attack")
class CentaurRangeAttack extends CustomAbility {
    GetAnimation() {
        return GameActivity.DOTA_ATTACK;
    }

    GetPlaybackRateOverride() {
        return 0.8;
    }

    GetCastingCrawl() {
        return 0;
    }

    OnAbilityPhaseStart() {
        if (super.OnAbilityPhaseStart()) {
            EFX(resources.castPoint.path, ParticleAttachment.ABSORIGIN_FOLLOW, this.caster, {
                relese: true
            });
            return true;
        }
        return false;
    }

    OnSpellStart() {
        const origin = this.caster.GetOrigin();
        const point = this.GetCursorPosition();
        const straightDirection = direction2D(origin, point);

        const directions = [
            straightDirection,
            RotatePoint(Vector(0, 0, 0), Vector(0, 0, 0).__add(straightDirection), -30).Normalized(),
            RotatePoint(Vector(0, 0, 0), Vector(0, 0, 0).__add(straightDirection), 30).Normalized()
        ];

        for (const direction of directions) {
            this.LaunchProjectile(direction);
        }

        EmitSoundOn("Hero_Centaur.DoubleEdge.TI9", this.caster);
    }

    LaunchProjectile(direction: Vector) {
        const origin = this.caster.GetOrigin();
        const projectileSpeed = this.GetSpecialValueFor("projectile_speed");

        if (fakeAbility(this)) {
            return;
        }

        this.ProjectileAttack({
            source: this.caster,
            effectName: resources.projectile.path,
            distance: this.GetCastRange(Vector(0, 0, 0), undefined),
            spawnOrigin: origin.__add(Vector(direction.x * 45, direction.y * 45, 96)),
            velocity: direction.__mul(projectileSpeed),
            groundOffset: 0,
            unitBehavior: ProjectileBehavior.NOTHING,
            unitTest: (unit, projectile) => !areUnitsAllied(projectile.getSource(), unit),
            onUnitHit: (unit, projectile) => {
                ApplyDamage({
                    victim: unit,
                    attacker: projectile.getSource(),
                    damage: this.GetSpecialValueFor("ability_damage"),
                    damage_type: DamageTypes.MAGICAL
                });
                this.PlayEffectsOnHit(unit);
            },
            onFinish: projectile => {
                this.PlayEffectsOnFinish(projectile.getPosition(), resources.impact.path);
            }
        });
    }

    PlayEffectsOnFinish(position: Vector, path: string) {
        EmitSoundOnLocationWithCaster(position, "Hero_Centaur.DoubleEdge", this.caster);
        EmitSoundOnLocationWithCaster(position, "Hero_Centaur.DoubleEdge.TI9_layer", this.caster);
        EmitSoundOnLocationWithCaster(position, "Hero_Centaur.Gore", this.caster);

        const particleId = ParticleManager.CreateParticle(path, ParticleAttachment.ABSORIGIN, this.caster);
        ParticleManager.SetParticleControl(particleId, 0, position);
        ParticleManager.SetParticleControl(particleId, 1, position);
        ParticleManager.SetParticleControl(particleId, 3, position);

        ParticleManager.ReleaseParticleIndex(particleId);
    }

    PlayEffectsOnHit(target: CDOTA_BaseNPC) {
        EmitSoundOn("Hero_Sniper.AssassinateDamage", this.caster);

        const particleId = ParticleManager.CreateParticle(resources.impactBlood.path, ParticleAttachment.ABSORIGIN_FOLLOW, target);
        ParticleManager.SetParticleControl(particleId, 0, target.GetAbsOrigin());
        ParticleManager.SetParticleControl(particleId, 1, target.GetAbsOrigin());
        ParticleManager.ReleaseParticleIndex(particleId);
    }
}
